### ISSUE TITLE
Installation Failure with Docker and Portainer

### ISSUE THREAD

**Issue by Dornski** [jessedp/pihole5-list-tool#10]
Hello, I use openmediavault with portainer and docker. And try to install the pihole adlist tool via docker exec -it pihole. I had to install python3.8, but then I get the error when I run the command. Is there a guide or help?

root@Pihole:/usr/bin/Python-3.8.2# pip install --upgrade setuptools
Requirement already up-to-date: setuptools in /usr/local/lib/python3.8/site-packages (47.1.1)

root@Pihole:/usr/bin/Python-3.8.2# sudo python3.8 -m pip install pihole5-list-tool
Collecting pihole5-list-tool
  Using cached pihole5_list_tool-0.4.2-py2.py3-none-any.whl (10 kB)
Collecting PyInquirer
  Using cached PyInquirer-1.0.3.tar.gz (27 kB)
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python3.8 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-vb0rpesu/PyInquirer/setup.py'"'"'; __file__='"'"'/tmp/pip-install-vb0rpesu/PyInquirer/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-kffb6gdg
         cwd: /tmp/pip-install-vb0rpesu/PyInquirer/
    Complete output (11 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/usr/local/lib/python3.8/site-packages/setuptools/__init__.py", line 19, in <module>
        from setuptools.dist import Distribution
      File "/usr/local/lib/python3.8/site-packages/setuptools/dist.py", line 34, in <module>
        from setuptools import windows_support
      File "/usr/local/lib/python3.8/site-packages/setuptools/windows_support.py", line 2, in <module>
        import ctypes
      File "/usr/local/lib/python3.8/ctypes/__init__.py", line 7, in <module>
        from _ctypes import Union, Structure, Array
    ModuleNotFoundError: No module named '_ctypes'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.

**Comment by MexSalem** [comment]
_ModuleNotFoundError: No module named '_ctypes'_

Installing `libffi-dev` and re-installing python3.8 fixed the problem for me. To cleanly build python3.8 `libffi-dev` is required or else later stuff will fail

If using RHEL/Fedora:
`yum install libffi-devel`

or

`sudo dnf install libffi-devel`

If using Debian/Ubuntu:

`sudo apt-get install libffi-dev`

Kudos goes to Steve E.:
https://stackoverflow.com/questions/27022373/python3-importerror-no-module-named-ctypes-when-using-value-from-module-mul

**Comment by Dornski** [comment]
Ok, thank you. It's work for me and the pihole5-list-tool starts.

But I have another mistake now. I start the tool and then I get the skipping error.

Skipping: /
Skipping: C
Skipping: o
Skipping: i
Skipping: n
Skipping: B
Skipping: l
Skipping: o
Skipping: c
Skipping: k
Skipping: e
Skipping: r
Skipping: L
Skipping: i
Skipping: s
Skipping: t
Skipping: s
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: _
Skipping: b
Skipping: r
Skipping: o
Skipping: w
Skipping: s
Skipping: e
Skipping: r
No valid urls found, try again

root@Pihole:/usr/bin/Python-3.8.2# ls /etc/pihole
GitHubVersions    dns-servers.conf  list.0.raw.githubusercontent.com.domains   list.3.s3.amazonaws.com.domains  localversions  migration_backup  setupVars.conf
adlists.list      gravity.db        list.1.mirror1.malwaredomains.com.domains  local.list                       logrotate      pihole-FTL.conf   setupVars.conf.update.bak
adlists.list.old  install.log       list.2.s3.amazonaws.com.domains            localbranches                    macvendor.db   pihole-FTL.db
root@Pihole:/usr/bin/Python-3.8.2#

**Comment by MexSalem** [comment]
Are you on Pihole 5 ? 
```
$ pihole -v       
  Pi-hole version is v5.0 (Latest: v5.0)
  AdminLTE version is v5.0 (Latest: v5.0)
  FTL version is v5.0 (Latest: v5.0)
```

Because your `ls /etc/pihole` seems different to mine (didn't see gravity.db):

```
$ ls /etc/pihole
GitHubVersions
adlists.list
black.list
dhcp.leases
dns-servers.conf
gravity.db
gravity.list
install.log
list.0.raw.githubusercontent.com.domains
list.1.raw.githubusercontent.com.domains
list.10.raw.githubusercontent.com.domains
list.11.raw.githubusercontent.com.domains
list.12.raw.githubusercontent.com.domains
list.13.raw.githubusercontent.com.domains
list.14.raw.githubusercontent.com.domains
list.15.raw.githubusercontent.com.domains
list.16.raw.githubusercontent.com.domains
list.17.raw.github.com.domains
...
```
If you're on Pi-hole version 5.0, update to latest version of `pihole5-list-tool`:
`$ sudo pip3 install pihole5-list-tool --upgrade`

If this fail, check your URL of "CoinBlockerLists/hosts_browser" or import:
https://gitlab.com/ZeroDot1/CoinBlockerLists/-/blob/master/hosts_browser

**Comment by Dornski** [comment]
Here is my output....

My Installation is - Openmediavault -- Docker -- Portainer -- Pihole:latest

docker exec -it pihole bash

root@Pihole:/# pihole -v
  Pi-hole version is v5.0 (Latest: v5.0)
  AdminLTE version is v5.0 (Latest: v5.0)
  FTL version is v5.0 (Latest: v5.0)

root@Pihole:/# sudo pip3 install pihole5-list-tool --upgrade
Requirement already up-to-date: pihole5-list-tool in /usr/local/lib/python3.8/si                           te-packages (0.4.3)
Requirement already satisfied, skipping upgrade: ansicolors in /usr/local/lib/py                           thon3.8/site-packages (from pihole5-list-tool) (1.1.8)
Requirement already satisfied, skipping upgrade: requests in /usr/local/lib/pyth                           on3.8/site-packages (from pihole5-list-tool) (2.23.0)
Requirement already satisfied, skipping upgrade: PyInquirer in /usr/local/lib/py                           thon3.8/site-packages (from pihole5-list-tool) (1.0.3)
Requirement already satisfied, skipping upgrade: idna<3,>=2.5 in /usr/local/lib/                           python3.8/site-packages (from requests->pihole5-list-tool) (2.9)
Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /usr/loca                           l/lib/python3.8/site-packages (from requests->pihole5-list-tool) (2020.4.5.1)
Requirement already satisfied, skipping upgrade: chardet<4,>=3.0.2 in /usr/local                           /lib/python3.8/site-packages (from requests->pihole5-list-tool) (3.0.4)
Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,                           >=1.21.1 in /usr/local/lib/python3.8/site-packages (from requests->pihole5-list-                           tool) (1.25.9)
Requirement already satisfied, skipping upgrade: prompt_toolkit==1.0.14 in /usr/                           local/lib/python3.8/site-packages (from PyInquirer->pihole5-list-tool) (1.0.14)
Requirement already satisfied, skipping upgrade: Pygments>=2.2.0 in /usr/local/l                           ib/python3.8/site-packages (from PyInquirer->pihole5-list-tool) (2.6.1)
Requirement already satisfied, skipping upgrade: regex>=2016.11.21 in /usr/local                           /lib/python3.8/site-packages (from PyInquirer->pihole5-list-tool) (2020.5.14)
Requirement already satisfied, skipping upgrade: six>=1.9.0 in /usr/local/lib/py                           thon3.8/site-packages (from prompt_toolkit==1.0.14->PyInquirer->pihole5-list-too                           l) (1.15.0)
Requirement already satisfied, skipping upgrade: wcwidth in /usr/local/lib/pytho                           n3.8/site-packages (from prompt_toolkit==1.0.14->PyInquirer->pihole5-list-tool)                            (0.2.3)

Here ist the gravitiy.db

root@Pihole:/# ls /etc/pihole
GitHubVersions    install.log                                local.list     migration_backup
adlists.list      list.0.raw.githubusercontent.com.domains   localbranches  pihole-FTL.conf
adlists.list.old  list.1.mirror1.malwaredomains.com.domains  localversions  pihole-FTL.db
dns-servers.conf  list.2.s3.amazonaws.com.domains            logrotate      setupVars.conf
**gravity.db**        list.3.s3.amazonaws.com.domains            macvendor.db   setupVars.conf.update.bak
root@Pihole:/#


    ┌──────────────────────────────────────────┐
    │       π-hole 5 list tool  v0.4.3         │
    └──────────────────────────────────────────┘
    https://github.com/jessedp/pihole5-list-tool

    Do not hit ENTER or Y if a step seems to hang!
    Use CTRL+C if you're sure it's hung and report it.

docker pihole image not found running, continuing...
? Gravity Db to Update:  /etc/pihole/gravity.db

**Comment by PatDal81** [comment]
Same error here, brand new installation (from one hour ago):

    ┌──────────────────────────────────────────┐
    │       π-hole 5 list tool  v0.4.3         │
    └──────────────────────────────────────────┘
    https://github.com/jessedp/pihole5-list-tool

    Do not hit ENTER or Y if a step seems to hang!
    Use CTRL+C if you're sure it's hung and report it.

docker pihole image not found running, continuing...
? Gravity Db to Update:  /etc/pihole/gravity.db
? Add Blacklists or Whitelists?  Blacklists
? Where are the block lists coming from?  Firebog | Ticked lists : Use where no one will be whitelisting falsely
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: r
Skipping: a
Skipping: w
Skipping: .
Skipping: g
Skipping: i
Skipping: t
Skipping: h
Skipping: u
Skipping: b
Skipping: u
Skipping: s
Skipping: e
Skipping: r
Skipping: c
Skipping: o
Skipping: n
Skipping: t
Skipping: e
Skipping: n
Skipping: t
Skipping: .
Skipping: c
Skipping: o
Skipping: m
Skipping: /
Skipping: P
Skipping: o
Skipping: l
Skipping: i
Skipping: s
Skipping: h
Skipping: F
Skipping: i
Skipping: l
Skipping: t
Skipping: e
Skipping: r
Skipping: s
Skipping: T
Skipping: e
Skipping: a
Skipping: m
Skipping: /
Skipping: K
Skipping: A
Skipping: D
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: /
Skipping: m
Skipping: a
Skipping: s
Skipping: t
Skipping: e
Skipping: r
Skipping: /
Skipping: K
Skipping: A
Skipping: D
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: _
Skipping: w
Skipping: i
Skipping: t
Skipping: h
Skipping: o
Skipping: u
Skipping: t
Skipping: _
Skipping: c
Skipping: o
Skipping: n
Skipping: t
Skipping: r
Skipping: o
Skipping: v
Skipping: e
Skipping: r
Skipping: s
Skipping: i
Skipping: e
Skipping: s
Skipping: .
Skipping: t
Skipping: x
Skipping: t
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: r
Skipping: a
Skipping: w
Skipping: .
Skipping: g
Skipping: i
Skipping: t
Skipping: h
Skipping: u
Skipping: b
Skipping: u
Skipping: s
Skipping: e
Skipping: r
Skipping: c
Skipping: o
Skipping: n
Skipping: t
Skipping: e
Skipping: n
Skipping: t
Skipping: .
Skipping: c
Skipping: o
Skipping: m
Skipping: /
Skipping: F
Skipping: a
Skipping: d
Skipping: e
Skipping: M
Skipping: i
Skipping: n
Skipping: d
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: .
Skipping: e
Skipping: x
Skipping: t
Skipping: r
Skipping: a
Skipping: s
Skipping: /
Skipping: m
Skipping: a
Skipping: s
Skipping: t
Skipping: e
Skipping: r
Skipping: /
Skipping: a
Skipping: d
Skipping: d
Skipping: .
Skipping: S
Skipping: p
Skipping: a
Skipping: m
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: v
Skipping: .
Skipping: f
Skipping: i
Skipping: r
Skipping: e
Skipping: b
Skipping: o
Skipping: g
Skipping: .
Skipping: n
Skipping: e
Skipping: t
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: /
Skipping: s
Skipping: t
Skipping: a
Skipping: t
Skipping: i
Skipping: c
Skipping: /
Skipping: w
Skipping: 3
Skipping: k
Skipping: b
Skipping: l
Skipping: .
Skipping: t
Skipping: x
Skipping: t
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: a
Skipping: d
Skipping: a
Skipping: w
Skipping: a
Skipping: y
Skipping: .
Skipping: o
Skipping: r
Skipping: g
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: .
Skipping: t
Skipping: x
Skipping: t
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: v
Skipping: .
Skipping: f
Skipping: i
Skipping: r
Skipping: e
Skipping: b
Skipping: o
Skipping: g
Skipping: .
Skipping: n
Skipping: e
Skipping: t
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: /
Skipping: A
Skipping: d
Skipping: g
Skipping: u
Skipping: a
Skipping: r
Skipping: d
Skipping: D
Skipping: N
Skipping: S
Skipping: .
Skipping: t
Skipping: x
Skipping: t
Skipping: h
Skipping: t
Skipping: t
Skipping: p
Skipping: s
Skipping: :
Skipping: /
Skipping: /
Skipping: v
Skipping: .
Skipping: f
Skipping: i
Skipping: r
Skipping: e
Skipping: b
Skipping: o
Skipping: g
Skipping: .
Skipping: n
Skipping: e
Skipping: t
Skipping: /
Skipping: h
Skipping: o
Skipping: s
Skipping: t
Skipping: s
Skipping: /
Skipping: A
Skipping: d

**Comment by jessedp** [comment]
Sigh. So the original issue was solved with @MexSalem 's solution.

In the middle of that I fixed #11 which broke the issue the "Skipping" problem.

That's fixed in 0.4.4 .
```yarn install pihole5-list-tool --upgrade```

Tests are almost done and will help prevent this.

**Comment by PatDal81** [comment]
I confirm the skipping problem is fixed with 0.4.4.  Thanks for the quick turnaround.
